src/jsutils/Future.ts

```diff
@@ -0,0 +1,20 @@
+import { invariant } from './invariant.js';
+import { isPromise } from './isPromise.js';
+import type { PromiseOrValue } from './PromiseOrValue.js';
+
+/** @internal */
+export class Future<T> {
```

Comment:
Can you explain how this class is used?

Reply:
Sure =>

I just pushed a version of this PR with many commits to show how when I extracted the incremental graph from the incremental publisher -- and made it mutation-less -- the end of the process (i.e., the second to last commit) introduced some extra ticks.

The Future class get rid of those extra ticks, as well as a whole bunch of others. Basically, `Future<T>` can be used wherever a `PromiseOrValue<T>` can be used => but a `Future<T>` will automatically update its internal value when the promise resolves, so that if you are doing some logic that is `if (isPromise(x)) { /* do A */ } else { /* do B */ }`, using a Future, you don't have to wait an extra tick if the promise has previously been resolved.

You can see now that almost all of the individual commits don't change tests, but now you can see which ones do, and hopefully that makes clearer why I introduced the `Future` class and what it's doing.

Reply:
Renamed `Future` to `BoxedPromiseOrValue`